### DISCUSSION TITLE
How to perform 'with' in some complex analysis instead of lm?

### DISCUSSION THREAD

**Discussion by Bigsealion** [amices/mice#332] [Impute, analyse and pool]
I want to perform SVM (package 'e1071') and randomForest(package 'randomForest') on imputed data, but I don't know how to apply this method by 'with' function.
Can any analysis be performed and pooled on imputed data? Or only some special method can using? What should I do if I want to apply and pool this method (like svm, randomforest, clustering analysis...) on imputed data? 
Thanks!

**Comment by gerkovink** [comment]
If you'd like to combine models without parameters or with different predictor sets across the imputed data sets, you'd need ensemble techniques. 

See e.g. [this vignette](https://www.gerkovink.com/miceVignettes/Combining_inferences/Combining_inferences.html) for an example that uses majority voting on stepwise selected models.

**Comment by gerkovink** [comment]
If your aim is to obtain predictions, it is more efficient to pool the predicted values and not pool the parameters/models first. See below for a `reprex` that demonstrates pooling predicted values. 

``` r
library(mice)     # Multiple Imputation
library(dplyr)    # Data manipulation
library(tidyr)    # Tidy data
library(magrittr) # Pipes
library(purrr)    # Functional programming - map()

set.seed(123) # Fix RNG seed

imp <- mice(boys, 
            maxit = 2,     # for reasons of brevity
            print = FALSE) # no iteration history

pred <- complete(imp, "all") %>% 
  map(lm, formula = hgt ~ age + wgt + tv) %>% # model
  map(predict) %>% # list of predicted values per imputed set
  Reduce("+", .) / imp$m # average corresponding list elements


cor(boys$hgt, pred, use = "pairwise.complete.obs")^2 #R-squared
#> [1] 0.9595445
```

<sup>Created on 2020-12-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by gerkovink** [comment]
`mice`'s function `pool()` is aimed at pooling parametric models cf. [Rubin's rules - see e.g. Section 2.3.2 in FIMD v.2](https://stefvanbuuren.name/fimd/sec-whyandwhen.html).

**Comment by Bigsealion** [comment]
Thank you for your answer! It's helpful! 
Beside, if I want to build a predict model that using X to predict y,  should I add y as a predictor of imputation? 
I'm worried that it will introduce information of label to predict model, but example in book 'Flexble Imputation of Missing Data' using all variable which contain interesting variable as predictor of imputaion. I want to know why the method in book is reasonable?
Thanks!

**Comment by gerkovink** [comment]
Yes, otherwise you'll render your prediction procedure uncongenial to the imputation model  - see e.g. [Meng (1994)](https://projecteuclid.org/euclid.ss/1177010269) and [Bartlett et al. (2015)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4513015/).

**Comment by gerkovink** [comment]
[FIMD Section 4.5](https://stefvanbuuren.name/fimd/sec-FCS.html) considers the relevant topics.

**Comment by stefvanbuuren** [comment]
Great discussion. Two remarks:

- OP: "Can any analysis be performed and pooled on imputed data?": **performed YES** (easy), **pooled YES** (but may be difficult, depending on the nature of the model)
- OP: "should I add y as a predictor of imputation?": **YES**, see <https://stefvanbuuren.name/fimd/sec-modelform.html#sec:predictors> for a recipe and reasons. Nevertheless, Little (1992) section 4.2 mentions a (not widely known) special case assuming MCAR where you should *not* include *y* and adapt case weights, so read that for the nitty-gritty.

**Comment by Bigsealion** [comment]
Thank you very much for your reply!
I will read this materials and try to solve my problem by mice!
Thanks!
